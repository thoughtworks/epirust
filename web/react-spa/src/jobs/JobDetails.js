/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";

import "./job-details.scss"
import { NavItem } from "../common/NavItem";
import { Redirect, Route, Switch } from "react-router-dom";
import GridPage from "../grid";
import PropTypes from 'prop-types';
import { TimeSeriesDeviation, TimeSeries } from "../time-series";

export const JobDetails = ({ simulationId, details }) => {
  const linkPrefix = `/jobs/${simulationId}`;

  const isGridEnabled = details && details.config.enable_citizen_state_messages,
    isFinished = details && details.status === 'finished';

  function renderTabs() {
    return (
      <div className="row">
        <div className="col-8">
          <ul className="nav nav-tabs nav-fill">
            <NavItem name="Time Series" linksTo={`${linkPrefix}/time-series`} activeOnExactMatch={true} />
            {isFinished && <NavItem name="Time Series Deviation" linksTo={`${linkPrefix}/time-series-deviation`} />}
            {isGridEnabled && <NavItem name="Grid" linksTo={`${linkPrefix}/grid`} />}
            <NavItem name="Config" linksTo={`${linkPrefix}/config`} />
          </ul>
        </div>
        <div className="col-4">
          <h4 className="simulation-id">Simulation: {simulationId}</h4>
        </div>
      </div>
    );
  }

  function renderContentForTab() {
    return (
      <Switch>

        <Route exact path={"/jobs/:id/time-series"}>
          <TimeSeries simulationId={simulationId} />
        </Route>

        <Route exact path={"/jobs/:id/time-series-deviation"}>
          <TimeSeriesDeviation simulationId={simulationId} />
        </Route>

        <Route exact path={"/jobs/:id/grid"}>
          {isGridEnabled ? <GridPage /> : <Redirect to={`/jobs/${simulationId}/time-series`} />}
        </Route>

        <Route exact path={"/jobs/:id/config"}>
          {details && <pre>{JSON.stringify(details.config, undefined, 4)}</pre>}
        </Route>
      </Switch>
    );
  }


  return (
    <div className="job-details" style={{ paddingTop: 4 }}>
      {renderTabs()}
      {renderContentForTab()}
    </div>
  )
};

JobDetails.propTypes = {
  simulationId: PropTypes.number.isRequired,
  details: PropTypes.object
};
