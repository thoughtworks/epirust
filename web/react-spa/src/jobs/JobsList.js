/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, { useEffect, useState } from "react";
import './jobs-list.scss'
import { Job } from "./Job";
import { JobDetails } from "./JobDetails";
import { Redirect, useParams } from 'react-router-dom';
import config from "../config";

export const JobsList = () => {
  const { id, view } = useParams();
  const [simulations, updateSimulations] = useState([]);

  useEffect(() => {
    fetch(`${config.API_HOST}/simulation/`)
      .then(res => res.json())
      .then(value => updateSimulations(value.reverse()))
  }, []);

  if (id && !view) {
    return (<Redirect to={`/jobs/${id}/time-series`} />);
  } else {
    const simulationDetails = simulations.find(s => s.simulation_id === parseInt(id));
    return (<div className="row jobs-list">
      <div className="col-3">
        <ul className="list-group scrollable">
          {simulations.map(s =>
            <Job
              key={s.simulation_id}
              simulationId={s.simulation_id}
              status={s.status} />
          )}
        </ul>
      </div>

      <div className="col-9 left-border scrollable">
        {simulationDetails && id && <JobDetails simulationId={parseInt(id)} details={simulationDetails} />}
      </div>

    </div>
    );
  }
}
