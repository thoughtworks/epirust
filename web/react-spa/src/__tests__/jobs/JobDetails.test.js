/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import { JobDetails } from "../../jobs/JobDetails";
import { render } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";

describe('Job Details', () => {
    it('should render the time series if the view is time-series', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/time-series']}>
            <JobDetails simulationId={123} details={{ config: {} }} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should render the config if the config is given', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/config']}>
            <JobDetails simulationId={123} details={{ config: { configKey: 'configValue' } }} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should not render the config if the config is not given', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/config']}>
            <JobDetails simulationId={123} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should render time-series if grid is disabled in config', () => {
        const simulationDetails = { config: { enable_citizen_state_messages: false } };
        const component = render(<MemoryRouter initialEntries={['/jobs/123/grid']}>
            <JobDetails simulationId={123} details={simulationDetails} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should render grid if grid is enabled in config', () => {
        const simulationDetails = { config: { enable_citizen_state_messages: true } };
        const component = render(<MemoryRouter initialEntries={['/jobs/123/grid']}>
            <JobDetails simulationId={123} details={simulationDetails} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    });

    it('should render time series mean comparison graph', () => {
        const component = render(<MemoryRouter initialEntries={['/jobs/123/time-series-deviation']}>
            <JobDetails simulationId={123} details={{ config: {} }} />
        </MemoryRouter>);

        expect(component.container).toMatchSnapshot()
    })
});