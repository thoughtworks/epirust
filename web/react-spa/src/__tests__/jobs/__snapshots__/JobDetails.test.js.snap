// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Details should not render the config if the config is not given 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 
          123
        </h4>
      </div>
    </div>
  </div>
</div>
`;

exports[`Job Details should not render time series mean comparison tab when a job is not finished 1`] = `
<DocumentFragment>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 123
        </h4>
      </div>
    </div>
    <div
      class="graph-wrap"
    >
      <div
        data-testid="visualization"
        id="vis"
      />
      <div
        class="graph-actions"
      >
        <p
          class="action"
        >
          Zoom: 
          <span
            class="control"
          >
            click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Pan: 
          <span
            class="control"
          >
            shift-click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Restore: 
          <span
            class="control"
          >
            double-click
          </span>
        </p>
        <button
          class="btn btn-secondary"
          disabled=""
        >
          Export graph
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Job Details should render grid if grid is enabled in config 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/grid"
            >
              Grid
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 
          123
        </h4>
      </div>
    </div>
    Loading
  </div>
</div>
`;

exports[`Job Details should render the config if the config is given 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 
          123
        </h4>
      </div>
    </div>
    <pre>
      {
    "configKey": "configValue"
}
    </pre>
  </div>
</div>
`;

exports[`Job Details should render the time series if the view is time-series 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 
          123
        </h4>
      </div>
    </div>
    <div
      class="graph-wrap"
    >
      <div
        data-testid="visualization"
        id="vis"
      />
      <div
        class="graph-actions"
      >
        <p
          class="action"
        >
          Zoom: 
          <span
            class="control"
          >
            click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Pan: 
          <span
            class="control"
          >
            shift-click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Restore: 
          <span
            class="control"
          >
            double-click
          </span>
        </p>
        <button
          class="btn btn-secondary"
          disabled=""
        >
          Export graph
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Job Details should render time series mean comparison graph tab when a job is finished 1`] = `
<DocumentFragment>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/time-series-deviation"
            >
              Time Series Deviation
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 123
        </h4>
      </div>
    </div>
    <div
      class="graph-wrap"
    >
      <div
        data-testid="visualization"
        id="vis"
      />
      <div
        class="graph-actions"
      >
        <p
          class="action"
        >
          Zoom: 
          <span
            class="control"
          >
            click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Pan: 
          <span
            class="control"
          >
            shift-click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Restore: 
          <span
            class="control"
          >
            double-click
          </span>
        </p>
        <button
          class="btn btn-secondary"
          disabled=""
        >
          Export graph
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Job Details should render time-series if grid is disabled in config 1`] = `
<div>
  <div
    class="job-details"
    style="padding-top: 4px;"
  >
    <div
      class="row"
    >
      <div
        class="col-8"
      >
        <ul
          class="nav nav-tabs nav-fill"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href="/jobs/123/time-series"
            >
              Time Series
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              class="nav-link "
              href="/jobs/123/config"
            >
              Config
            </a>
          </li>
        </ul>
      </div>
      <div
        class="col-4"
      >
        <h4
          class="simulation-id"
        >
          Simulation: 
          123
        </h4>
      </div>
    </div>
    <div
      class="graph-wrap"
    >
      <div
        data-testid="visualization"
        id="vis"
      />
      <div
        class="graph-actions"
      >
        <p
          class="action"
        >
          Zoom: 
          <span
            class="control"
          >
            click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Pan: 
          <span
            class="control"
          >
            shift-click-drag
          </span>
        </p>
        <p
          class="action"
        >
          Restore: 
          <span
            class="control"
          >
            double-click
          </span>
        </p>
        <button
          class="btn btn-secondary"
          disabled=""
        >
          Export graph
        </button>
      </div>
    </div>
  </div>
</div>
`;
